main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -28,25 +30,27 @@ object Safety {
 
   /** Checks the safety and well-formedness of `root`. */
   def run(root: Root)(implicit flix: Flix): (Root, List[SafetyError]) = flix.phaseNew("Safety") {
-    val classSigErrs = ParOps.parMap(root.traits.values.flatMap(_.sigs))(visitSig).flatten
-    val defErrs = ParOps.parMap(root.defs.values)(visitDef).flatten
-    val instanceDefErrs = ParOps.parMap(TypedAstOps.instanceDefsOf(root))(visitDef).flatten
-    val sigErrs = ParOps.parMap(root.sigs.values)(visitSig).flatten
-    val errors = classSigErrs ++ defErrs ++ instanceDefErrs ++ sigErrs
+    implicit val sctx: SharedContext = SharedContext.mk()
+    ParOps.parMap(root.traits.values.flatMap(_.sigs))(visitSig)
+    ParOps.parMap(root.defs.values)(visitDef)
+    ParOps.parMap(TypedAstOps.instanceDefsOf(root))(visitDef)
+    ParOps.parMap(root.sigs.values)(visitSig)
 
-    (root, errors.toList)
+    (root, sctx.errors.asScala.toList)
   }
 
   /** Checks the safety and well-formedness of `sig`. */
-  private def visitSig(sig: Sig)(implicit flix: Flix): List[SafetyError] = {
+  private def visitSig(sig: Sig)(implicit flix: Flix, sctx: SharedContext): Sig = {
     val renv = RigidityEnv.ofRigidVars(sig.spec.tparams.map(_.sym))
-    sig.exp.map(visitExp(_)(inTryCatch = false, renv, flix)).getOrElse(Nil)
+    sig.exp.foreach(visitExp(_)(inTryCatch = false, renv, flix, sctx))
+    sig
   }
 
   /** Checks the safety and well-formedness of `defn`. */
-  private def visitDef(defn: Def)(implicit flix: Flix): List[SafetyError] = {
+  private def visitDef(defn: Def)(implicit flix: Flix, sctx: SharedContext): Def = {
```

Comment:
We always order implicit parameters from most specific to least specific. The upshot is that the Flix argument should always be last.